### PR TITLE
Sever `PreservationTable` inheritance and add utility methods

### PR DESCRIPTION
**Pull request by JosiahWI**
This adds the following methods:

* find(Dir const &)
* acquire(Dir const&, CacheKey const&)
* release(Dir const&)

The latter two are used for reference counting readers of documents so that the documents will be preserved while they're open for reading.

### CODE REVIEW

File: src/iocore/cache/CacheEvacuateDocVC.cc [hunk 1]
```diff
@@ -57,88 +57,84 @@ CacheEvacuateDocVC::evacuateDocDone(int /* event ATS_UNUSED */, Event * /* e ATS
...
+    // only overwrite the entry in the directory if its not a head.
+    if (!dir_head(&this->overwrite_dir)) {
+      // find the earliest key
+      EvacuationKey *evac = &b->evac_frags;
+      for (; evac && !(evac->key == doc->key); evac = evac->link.next) {
+        ;
+      }
+      ink_assert(evac);
+      if (!evac) {
+        goto RET;
```
Comment (masaori335) [inline src/iocore/cache/CacheEvacuateDocVC.cc:77]: How about simply returning from here?

If calling `free_CacheEvacuateDocVC(this)` is the matter, setting defer by `ts::PostScript` in the beginning is another apporach.
```
ts::PostScript defer_free([&](){ free_CacheEvacuateDocVC(this); });
```

**Review by masaori335 — CHANGES_REQUESTED** [review]
Overall, I love this change, thank you! Please take a look at my comments.

File: src/iocore/cache/CacheEvacuateDocVC.cc (hunk 1 above)
Comment (JosiahWI) [reply src/iocore/cache/CacheEvacuateDocVC.cc:77]: In other branches we don't free it. Maybe the cleanest thing for now is to simply return as you say and also duplicate the call to free.

File: src/iocore/cache/StripeSM.cc [hunk 2]
```diff
@@ -1295,13 +1261,13 @@ evacuate_fragments(CacheKey *key, CacheKey *earliest_key, int force, StripeSM *s
     if (dir_head(&dir)) {
       continue;
     }
-    EvacuationBlock *b = evacuation_block_exists(&dir, stripe);
```
[the remark this answers is not in the document]
Comment (JosiahWI) [reply src/iocore/cache/StripeSM.cc:1298]: Good catch!

**Review by masaori335 — APPROVED** [review]
Looks good!

**Pull request merged**